cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1214,6 +1215,24 @@ Status SetObjectMetadata(const std::shared_ptr<const KeyValueMetadata>& metadata
   return Status::OK();
 }
 
+bool IsDirectory(std::string_view key, const S3Model::HeadObjectResult& result) {
+  // If it has a non-zero length, it's a regular file
+  if (result.GetContentLength() > 0) {
+    return false;
+  }
```

Comment:
I don't think it's even possible for all `arrow::FileSystem` functions to deal with `/`-ending paths that are sometimes considered to be a "file". It's better to simply say "we do not support reading files that end in a `/` in `arrow::FileSystem` implementations".

Reply:
This is the information that comes back from S3, not what we present to the user.

Reply:
Ah, sorry, I misread your comment. Well, we were originally cautious on this. It's also not obvious to me that AWS or other S3 implementations would always return an ending slash for pseudo-"directories". It seems to depend on _how_ the directory was created.

Reply:
I know, but we can limit the Arrow implementation in saying that it's unable to reason about files ending with a `/`. Directories without a `/` should be OK. It won't read and won't create files at keys ending with a `/`. Without this, it's really hard to offer filesystem operations that behave as expected.

Reply:
To be clear, this PR shouldn't change anything in that regard. It just adds the Content-Type heuristic, which seems desirable since it's shared with other tools.

Reply:
I'm closing this thread as the discussion is the same as in the other thread.